src/lib/block_producer/block_producer.ml

```diff
@@ -1097,10 +1096,26 @@ let run ~context:(module Context : CONTEXT) ~vrf_evaluator ~prover ~verifier
             let new_global_slot = epoch_data_for_vrf.global_slot in
             let log_if_slot_diff_is_less_than =
               let current_global_slot =
-                Consensus.Data.Consensus_time.(
-                  to_global_slot
-                    (of_time_exn ~constants:consensus_constants
-                       (Block_time.now time_controller) ))
+                let open Mina_numbers in
+                let global_slots_since_hard_fork =
+                  let global_slot_since_hard_fork =
+                    Consensus.Data.Consensus_time.(
+                      to_global_slot
+                        (of_time_exn ~constants:consensus_constants
+                           (Block_time.now time_controller) ))
+                  in
+                  Option.value_exn
+                  @@ Global_slot_since_hard_fork.(
+                       diff global_slot_since_hard_fork zero)
+                in
+                let open Mina_numbers in
+                Global_slot_since_genesis.add
+                  ( match constraint_constants.fork with
+                  | None ->
+                      Global_slot_since_genesis.zero
+                  | Some { global_slot_since_genesis; _ } ->
+                      global_slot_since_genesis )
+                  global_slots_since_hard_fork
```

Comment:
We seem to be doing this sort of computation in mutliple places in this PR. Can we turn this into a utility function at the global slot module layer? I don't want us getting this conversion wrong in some contexts.